npy_array/dynamic_array.h

```diff
@@ -0,0 +1,463 @@
+#ifndef NPY_ARRAY_DYNAMIC_ARRAY_H_
+#define NPY_ARRAY_DYNAMIC_ARRAY_H_
+
+#include <cstdint>
+#include <vector>
+
+#include "absl/types/span.h"
+#include "array/array.h"
+#include "npy_array/compile_time_loop.h"
+#include "npy_array/data_type.h"
+#include "npy_array/half.h"
+
+namespace npy_array {
+
+class DynamicShape {
+ public:
+  // DynamicShape intentionally lacks a default constructor because it would be
+  // the same as a scalar, which is uncommnon and surprising. It also makes the
+  // expression `DynamicShape({})` ambiguous (it should explicitly create a
+  // scalar).
+
+  // Constructs a DynamicShape with the given extents, automatically deducing:
+  // - mins to be all 0.
+  // - strides to be the product of all previous extents, starting with 1
+  // (i.e., a planar layout).
+  explicit DynamicShape(absl::Span<const int64_t> extents);
+
+  // Constructs a DynamicShape with the given mins, extents, and strides.
+  // TODO(jiawen): take in a Span<nda::dim<>>?
+  // TODO(jiawen): Test for unresolved strides?
+  DynamicShape(absl::Span<const int64_t> mins,
```

Comment:
@dsharlet @fbleibel I finally wrote a fully dynamic array class - it's what you expect.

I don't know if you have opinions on the design. It's fairly minimal.

One question - do you prefer `size_t` instead of `int64_t` for extents here? It's slightly less typing for interop because `NpyShape` and `array` both use `ssize_t`. @yyuting FYI

Reply:
Re: naming/casing: I ended up going with the compromise.

Re: `int64_t` vs `ptrdiff_t` vs `ssize_t`: ugh. I'll file an issue. Punt this until I do the pybind.